src/NServiceBus.Core/Unicast/DisposableEndpointInstance.cs

```diff
@@ -0,0 +1,56 @@
+#if NETCOREAPP
+namespace NServiceBus
+{
+    using System;
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    /// <summary>
+    /// Contains extension methods to create a <see cref="IDisposableEndpointInstance"/>.
+    /// </summary>
+    public static class DisposableEndpointExtensions
+    {
+        /// <summary>
+        /// Use this method to turn a starting endpoint instance into a <see cref="IDisposableEndpointInstance"/> that will stop the endpoint automatically on dispose.
+        /// If the endpoint is stopped during disposal, it will use a completed <see cref="CancellationToken"/>, indicating that it expects immediate shutdown.
+        /// </summary>
+#pragma warning disable PS0018 // A task-returning method should have a CancellationToken parameter unless it has a parameter implementing ICancellableContext
+        public static async Task<IDisposableEndpointInstance> AsDisposable(this Task<IEndpointInstance> startTask)
+#pragma warning restore PS0018 // A task-returning method should have a CancellationToken parameter unless it has a parameter implementing ICancellableContext
+        {
+            var runningInstance = await startTask.ConfigureAwait(false);
+            return new DisposableEndpointInstance(runningInstance);
+        }
+    }
+
+    /// <summary>
+    /// an <see cref="IEndpointInstance"/> that implements <see cref="IAsyncDisposable"/>. The endpoint performs a graceful shutdown when being disposed.
+    /// </summary>
+    public interface IDisposableEndpointInstance : IEndpointInstance, IAsyncDisposable
+    {
+    }
+
+    class DisposableEndpointInstance : IDisposableEndpointInstance
+    {
+        IEndpointInstance endpointInstance;
+
+        public DisposableEndpointInstance(IEndpointInstance endpointInstance) => this.endpointInstance = endpointInstance;
+
+        ValueTask IAsyncDisposable.DisposeAsync() => new ValueTask(endpointInstance.Stop(CancellationToken.None));
```

Comment:
Seeing this makes me wonder is we want this. Would it maybe make sense to pass this via `AsDisposable`?

Not doing this seems more that the user would still need to do 

```c#
var hostCancellationToken = ....;
await using var endpoint = await Endpoint.Start(config, hostCancellationToken ).AsDisposable();
await endpoint.Stop(hostCancellationToken);
```

So then you would get:


```c#
var hostCancellationToken = ....;
await using var endpoint = await Endpoint
    .Start(config, hostCancellationToken)
    .AsDisposable(hostCancellationToken);
```

Reply:
yeah currently if you want to explicitly control shutdown behavior via cancellation tokens, you need to call `Stop` manually. I think that's "good enough" since `AsDisposable` isn't intended to fully replace `Stop` but to just provide a convenient API for most cases where you don't care about passing cancellation tokens when stopping an endpoint. Your suggested approach would work if you are able to produce a meaningful token for shutdown already at startup time.